source/simulator/melt.cc

```diff
@@ -116,32 +119,35 @@ namespace aspect
           const bool use_tensor = (stress_strain_director != dealii::identity_tensor<dim> ());
 
           for (unsigned int i=0; i<dofs_per_cell; ++i)
-            for (unsigned int j=0; j<dofs_per_cell; ++j)
-              if (fe.system_to_component_index(i).first
-                  ==
-                  fe.system_to_component_index(j).first)
-                data.local_matrix(i,j) += ((use_tensor ?
-                                            eta * (scratch.grads_phi_u[i] * stress_strain_director * scratch.grads_phi_u[j])
-                                            :
-                                            eta * (scratch.grads_phi_u[i] * scratch.grads_phi_u[j]))
-                                           +
-                                           (1./eta *
-                                            pressure_scaling *
-                                            pressure_scaling)
-                                           * scratch.phi_p[i] * scratch.phi_p[j]
-                                           +
-                                           (K_D *
-                                            pressure_scaling *
-                                            pressure_scaling) *
-                                           scratch.grad_phi_p[i] *
-                                           scratch.grad_phi_p[j]
-                                           +
-                                           (1./eta + 1./viscosity_c) *
-                                           pressure_scaling *
-                                           pressure_scaling *
-                                           (scratch.phi_p_c[i] * scratch.phi_p_c[j])
-                                          )
-                                          * scratch.finite_element_values.JxW(q);
+            if (introspection.is_stokes_component(fe.system_to_component_index(i).first) ||
+                (fe.system_to_component_index(i).first == p_f_component_index)
```

Comment:
you are referencing the same variable 4-5 times. Maybe it makes sense to do
```
const unsigned int component_i = fe.system_to_component_index(i).first;
```
?